inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/permute/permute_kernel_tile_8x8_4x4.cpp

```diff
@@ -0,0 +1,330 @@
+                {
+                    lws[dim] = std::max(lws[dim], lws0);
+                }
+                if (divider <= max_num_work_items)
+                {
+                    lws[dim] = std::max(lws[dim], divider);
+                }
+            }
+        }
+        max_num_work_items /= lws[dim];
+    }
+    return lws;
+}
+
+CommonDispatchData PermuteKernel_tile_8x8_4x4::SetDefault(const permute_params& params, const size_t tile_size) const {
+    CommonDispatchData dispatchData;
+    const auto& in =  params.inputs[0];
+    switch (in.GetLayout()) {
+        case DataLayout::bfyx:
+            dispatchData.gws = {CEIL_DIV(in.X().v , tile_size), in.Y().v, CEIL_DIV(in.Feature().v, tile_size) * in.Batch().v};
+            break;
+        case DataLayout::bfzyx:
+            dispatchData.gws = {CEIL_DIV(in.X().v , tile_size), in.Y().v * in.Z().v, CEIL_DIV(in.Feature().v, tile_size) * in.Batch().v};
+            break;
+        case DataLayout::bfwzyx:
+            dispatchData.gws = {CEIL_DIV(in.X().v , tile_size), in.Y().v * in.Z().v * in.W().v, CEIL_DIV(in.Feature().v, tile_size) * in.Batch().v};
+            break;
+        default:
+            throw std::runtime_error("Unsupported combination\n");
+            break;
+    }
+    dispatchData.lws = GetBestLwsFromGws(params, dispatchData.gws, tile_size);
+    return dispatchData;
+}
+
+bool PermuteKernel_tile_8x8_4x4::Validate(const Params& p, const optional_params& o) const {
+    if (p.GetType() != KernelType::PERMUTE || o.GetType() != KernelType::PERMUTE) {
+        return false;
+    }
+
+    const permute_params& params = static_cast<const permute_params&>(p);
+
+    for (auto& fused_op : params.fused_ops) {
+        if (!IsFusedPrimitiveSupported(fused_op))
+            return false;
+    }
+
+    if (params.output.PitchesDifferFromLogicalDims() || params.inputs[0].PitchesDifferFromLogicalDims()) {
+        return false;
+    }
+
+    return true;
+}
+
+KernelsData PermuteKernel_tile_8x8_4x4::GetKernelsData(const Params& params, const optional_params& options) const {
+    if (!Validate(params, options)) {
+        return {};
+    }
+
+    KernelData kd = KernelData::Default<permute_params>(params);
+    permute_params& newParams = *static_cast<permute_params*>(kd.params.get());
+    size_t tile_size = DEFAULT_TILE_SIZE;
+    // supports 4x4 or 8x8 tiling
+    if (newParams.inputs[0].X().v < DEFAULT_TILE_SIZE || newParams.inputs[0].Feature().v < DEFAULT_TILE_SIZE)
+        tile_size = MIN_TILE_SIZE;
+
+    auto dispatchData = SetDefault(newParams, tile_size);
+    auto cldnn_jit = GetJitConstants(newParams, dispatchData, tile_size);
+    auto entry_point = GetEntryPoint(kernelName, newParams.layerID, options);
+    std::string jit = CreateJit(kernelName, cldnn_jit, entry_point);
+    auto& kernel = kd.kernels[0];
+    FillCLKernelData(kernel, dispatchData, params.engineInfo, kernelName, jit, entry_point, "", false, false, 1, GetFusedPrimitiveInputsCount(params));
```

Comment:
I'd propose to create permute_bass class and move some common code there.

Reply:
Creted permute_kernel_base.h and extracted common codes.